db-engine/src/main/java/com/splicemachine/db/impl/jdbc/EmbedPreparedStatement.java

```diff
@@ -1120,20 +1120,21 @@ public void clearParameters() throws SQLException {
                 //Second check - if the statement was revalidated since last getMetaData call,
                 //then gcDuringGetMetaData wouldn't match with current generated class name
 
-                GeneratedClass currAc;
-                ResultDescription resd;
+                GeneratedClass currAc = null;
+                ResultDescription resd = null;
 
-                synchronized(execp) {
-                    // DERBY-3823 Some other thread may be repreparing
-                    do {
-                        while (!execp.upToDate()) {
+                // DERBY-3823 Some other thread may be repreparing
+                do {
+                    synchronized (execp) {
+                        if (!execp.upToDate()) {
                             execp.rePrepare(lcc);
                         }
-
-                        currAc = execp.getActivationClass();
-                        resd = execp.getResultDescription();
-                    } while (currAc == null);
-                }
+                        else {
+                            currAc = execp.getActivationClass();
+                            resd = execp.getResultDescription();
+                        }
+                    }
```

Comment:
Should we put a short Thread.sleep() time in this loop so it's not consuming CPU if the blocking thread takes a long time:  
if (currAc == null)
    Thread.sleep(x);

Reply:
All operations are still executed in a synchronized block so it's not busy waiting.
There still is a chance it continuously fails to rePrepare execp due to broken logic somewhere else (not because other threads updating execp are blocked -- that's fixed here). In that case, I'd prefer it to burn CPU so that it's easily visible (I used this symptom to quickly check whether my benchmark runs were affected in AWS) rather than be hidden behind a sleep, which wouldn't help with that broken logic anyway.

Reply:
OK